File: sparc-backend/algorithms/dijkstra.js
/**
 * Dijkstra's Algorithm implementation
 * Computes the shortest paths from a source node to all other nodes.
 * 
 * @param {Object} graph - { nodes: number, edges: Array<{source, target, weight}> }
 * @param {number} source - The source node index
 * @returns {Object} { distances, path } or simple result
 */
function dijkstra(graph, source) {
  const { nodes, edges } = graph;
  const numNodes = Array.isArray(nodes) ? nodes.length : nodes;
  
  // Initialize distances
  const distances = Array(numNodes).fill(Infinity);
  distances[source] = 0;

  // Track the visited nodes
  const visited = Array(numNodes).fill(false);
  const previous = Array(numNodes).fill(null);

  // Build adjacency list for faster lookup
  const adj = Array.from({ length: numNodes }, () => []);
  for (const edge of edges) {
    adj[edge.source].push({ target: edge.target, weight: edge.weight });
    // Assuming directed graph. If undirected, add the reverse edge here.
    // The frontend payload usually implies directed based on standard datasets.
  }

  const startTime = process.hrtime.bigint();

  for (let i = 0; i < numNodes; i++) {
    // Find the unvisited node with the smallest distance
    let u = -1;
    let minDistance = Infinity;
    
    for (let j = 0; j < numNodes; j++) {
      if (!visited[j] && distances[j] < minDistance) {
        minDistance = distances[j];
        u = j;
      }
    }

    if (u === -1) break; // All remaining vertices are inaccessible

    visited[u] = true;

    // Update distances of adjacent nodes
    for (const neighbor of adj[u]) {
      const v = neighbor.target;
      const weight = neighbor.weight;

      if (!visited[v] && distances[u] !== Infinity && distances[u] + weight < distances[v]) {
        distances[v] = distances[u] + weight;
        previous[v] = u;
      }
    }
  }

  const endTime = process.hrtime.bigint();
  const timeTakenNs = Number(endTime - startTime);

  return {
    algorithm: 'DIJKSTRA',
    executionTimeMs: timeTakenNs / 1_000_000,
    distances,
    previous,
    nodesVisited: numNodes,
    relaxations: edges.length,
    hasNegativeCycle: false // Dijkstra doesn't handle/detect negative cycles properly
  };
}

module.exports = dijkstra;

File: sparc-backend/Dijkstra.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.List;
import java.util.Collections;

public class Dijkstra {
    static class NodeDist implements Comparable<NodeDist> {
        int id; double dist;
        NodeDist(int i, double d) { id = i; dist = d; }
        public int compareTo(NodeDist o) { return Double.compare(dist, o.dist); }
    }

    public static AlgorithmResult run(Graph graph, int source, Integer target) {
        long startTime = System.nanoTime();
        int numNodes = graph.nodes.size();
        
        double[] dist = new double[numNodes];
        Arrays.fill(dist, Double.POSITIVE_INFINITY);
        dist[source] = 0;
        
        Integer[] prev = new Integer[numNodes];
        boolean[] visited = new boolean[numNodes];
        
        PriorityQueue<NodeDist> pq = new PriorityQueue<>();
        pq.add(new NodeDist(source, 0));
        
        int relaxations = 0;
        int nodesVisited = 0;
        
        List<List<Graph.Edge>> adj = new ArrayList<>();
        for (int i = 0; i < numNodes; i++) adj.add(new ArrayList<>());
        for (Graph.Edge e : graph.edges) {
            adj.get(e.source).add(e);
            if (!graph.directed) adj.get(e.target).add(new Graph.Edge(e.target, e.source, e.weight));
        }

        while (!pq.isEmpty()) {
            NodeDist cur = pq.poll();
            if (visited[cur.id]) continue;
            visited[cur.id] = true;
            nodesVisited++;
            
            if (target != null && cur.id == target) break;

            for (Graph.Edge edge : adj.get(cur.id)) {
                relaxations++;
                double newDist = dist[cur.id] + edge.weight;
                if (newDist < dist[edge.target]) {
                    dist[edge.target] = newDist;
                    prev[edge.target] = cur.id;
                    pq.add(new NodeDist(edge.target, newDist));
                }
            }
        }

        long endTime = System.nanoTime();

        AlgorithmResult res = new AlgorithmResult();
        res.algorithm = "DIJKSTRA";
        res.executionTimeMs = (endTime - startTime) / 1_000_000.0;
        
        Object[] outDist = new Object[numNodes];
        for (int i = 0; i < numNodes; i++) outDist[i] = dist[i] == Double.POSITIVE_INFINITY ? null : (int)dist[i];
        res.distances = outDist;

        List<Integer> path = new ArrayList<>();
        if (target != null && prev[target] != null) {
            Integer curr = target;
            while (curr != null) {
                path.add(curr);
                curr = prev[curr];
            }
            Collections.reverse(path);
        } else if (target != null && source == target) {
            path.add(source);
        }
        res.path = path;
        res.nodesVisited = nodesVisited;
        res.relaxations = relaxations;

        return res;
    }
}
